### DISCUSSION TITLE
Model-level optional field validation

### DISCUSSION THREAD

**Discussion by khalil-omer** [prisma/prisma#16678] [Q&A] [answered]
Zod-prisma outputs z.objects for each of our Prisma models, and it allows us to customize the validation of each Prisma field. All we have to do is call `parse` before calling Prisma insert/upsert/etc., and we end up with great input validation for Prisma which stays up to date with schema changes. 

However, zod-prisma does not help us validate optional fields being conditionally required. Meaning, we have a few Prisma models where  `fieldA || fieldB` is required, but not both. Zod-prisma only allows adding custom validation schemas for specific fields, not for a model as a whole, so we are stuck with our Prisma client allowing a very invalid state.

Prisma Client Extensions seems like a great feature, but looking at the docs for extensions, I don't see a way to add any custom validation logic to the build-in methods, so I don't think it addresses this.

Consider the following example:

```

model Example {
  id Int
  foo String?
  bar String?
  type TypeEnum
  ...
}

enum TypeEnum {
  FOO
  BAR
}
```

If type = FOO, then the field `foo` should be required and the field `bar` should be null. However, we can't validate that at the database level or with the zod-prisma package. 

We can create simple TS functions to do this of course, but would prefer to have that included in our Prisma Client (or at least in the Zod schemas generated from our Prisma Client). Any ideas?

**Comment by nurul3101** [comment, marked as answer]
Hello @khalil-omer 👋 

Looking at your use case, wouldn't this [example](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/query#add-or-modify-a-custom-query-for-a-specific-operation-in-a-specific-model) fit in?
You can check in the where clause if foo or bar is present or not depending upon the type value and only proceed if the condition satisfies?

**Reply by khalil-omer** [reply]
Hey thanks @nurul3101. I didn't realize you could use Client Extensions to modify the standard operations.  Yes, I think this will work. The example in that article is for a find operation, but I will try the same for create/update and see how that works.

**Reply by nurul3101** [reply]
Sure, Let me know if you run into any issues 👍

**Comment by christianalares** [comment]
I'd like to bump this one as i suspect the provided link has changed since it was posted. I'm also looking for this behavior. Here is my example:

```
// schema.prisma

enum ThreadEventType {
  MESSAGE
  CLOSED_THREAD
}

model Thread {
  id            Int
  subject       String
  thread_events ThreadEvent[]
}

model ThreadEvent {
  id         Int
  thread_id  Int
  thread     @relation(fields: [thread_id], references: [id])
  type       ThreadEventType
  message_id String?
  message    Message?        @relation(fields: [message_id], references: [id])
}
``` 

```typescript
const thread = await prisma.thread.findMany({
  select: {
    id: true,
    thread_events: {
      select: {
        id: true,
        type: true,
        message: {
          select: {
            id: true,
            body: true,
          },
        },
      },
    },
  },
})

// ...

thread.thread_events.map(event => {
  if (event.type === 'MESSAGE') {
  // Would like the type inference that event has a message with a body since the type is `MESSAGE`
    return `Message: ${event.message.body}`
  }

  if (event.type === 'CLOSED_THREAD') {
    return 'Thread closed'
  }
})
```

I would like to have the same type safety to the other operations. Examples:
```typescript
prisma.thread_events.create({
  data: {
    type: 'MESSAGE', // ❌ Error because I also need to pass message
  },
})

prisma.thread_events.create({
  data: {
    type: 'CLOSED_THREAD', // ✅ Ok because you can't pass message
  },
})

prisma.thread_events.create({
  data: {
    type: 'CLOSED_THREAD', // ❌ Error because this type can't have message
    message: {
      create: {
        body: 'Hello',
      },
    },
  },
})

prisma.thread_events.create({
  data: {
    type: 'MESSAGE', // ✅ Ok, because this type can have message
    message: {
      create: {
        body: 'Hello',
      },
    },
  },
})
```

Is this behavior even possible or should I reconsider the models here?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
